kernel/src/engine_data.rs

```diff
@@ -182,26 +184,48 @@ impl<'a> TypedGetData<'a, HashMap<String, String>> for dyn GetData<'a> + '_ {
     }
 }
 
-/// A `DataVisitor` can be called back to visit extracted data. Aside from calling
-/// [`DataVisitor::visit`] on the visitor passed to [`EngineData::extract`], engines do
+/// A `RowVisitor` can be called back to visit extracted data. Aside from calling
+/// [`RowVisitorBase::visit`] on the visitor passed to [`EngineData::visit_rows`], engines do
 /// not need to worry about this trait.
-pub trait DataVisitor {
+pub trait RowVisitorBase {
+    /// The names and types of leaf fields this visitor accesses. The `EngineData` being visited
+    /// validates these types when extracting column getters, and [`RowVisitorBase::visit`] will
+    /// receive one getter for each selected field, in the requested order. The column names are
+    /// used by [`RowVisitor::visit_rows_of`] to select fields from a "typical" `EngineData`; callers
+    /// whose engine data has different column names can manually invoke [`EngineData::visit_rows`].
+    fn selected_column_names_and_types(&self) -> (&'static [ColumnName], &'static [DataType]);
+
     /// Have the visitor visit the data. This will be called on a visitor passed to
-    /// [`EngineData::extract`]. For each leaf in the schema that was passed to `extract` a "getter"
-    /// of type [`GetData`] will be present. This can be used to actually get at the data for each
-    /// row. You can `use` the `TypedGetData` trait if you want to have a way to extract typed data
-    /// that will fail if the "getter" is for an unexpected type.  The data in `getters` should not
-    /// be assumed to live beyond the call to this funtion (i.e. it should be copied if needed)
+    /// [`EngineData::visit_rows`]. For each leaf in the schema that was passed to `extract` a
+    /// "getter" of type [`GetData`] will be present. This can be used to actually get at the data
+    /// for each row. You can `use` the `TypedGetData` trait if you want to have a way to extract
+    /// typed data that will fail if the "getter" is for an unexpected type.  The data in `getters`
+    /// does not outlive the call to this funtion (i.e. it should be copied if needed).
     fn visit<'a>(&mut self, row_count: usize, getters: &[&'a dyn GetData<'a>]) -> DeltaResult<()>;
 }
 
+/// Helper trait that simplifies use of [`RowVisitor`] interface, implemented automatically for all
+/// types that implement [`RowVisitor`].
+pub trait RowVisitor: RowVisitorBase {
+    /// Visit the rows of an [`EngineData`], selecting the leaf column names given by
+    /// [`RowVisitorBase::selected_column_names_and_types`]. This is a thin wrapper around
+    /// [`EngineData::visit_rows`] which in turn will eventually invoke [`RowVisitorBase::visit`].
+    fn visit_rows_of(&mut self, data: &dyn EngineData) -> DeltaResult<()>;
+}
```

Comment:
could this just be a provided method of `RowVisitorBase`? (and then just rename `RowVisitorBase` to `RowVisitor`?)

Reply:
```
error[E0277]: the size for values of type `Self` cannot be known at compilation time
   --> kernel/src/engine_data.rs:210:67
    |
210 |         data.visit_rows(self.selected_column_names_and_types().0, self)
    |                                                                   ^^^^ doesn't have a size known at compile-time
    |
    = note: required for the cast from `&mut Self` to `&mut dyn engine_data::RowVisitor`
help: consider further restricting `Self`
    |
209 |     fn visit_rows_of(&mut self, data: &dyn EngineData) -> DeltaResult<()> where Self: std::marker::Sized {
    |                                                                           ++++++++++++++++++++++++++++++
```
Oho! I had originally updated the trait to be `trait RowVisitor: Sized`, which apparently made it not object-safe (can't cast to any kind of `& dyn`). But following the actual clippy recommendation compiled just fine.

Reply:
ahh i see, you just need to make `visit_rows_of` require `Self: Send` - cool thanks!